server/src/main/java/keywhiz/service/providers/ClientAuthFactory.java

```diff
@@ -79,51 +79,25 @@ public ClientAuthFactory(ClientDAOFactory clientDAOFactory, KeywhizConfig keywhi
   }
 
   public Client provide(ContainerRequest request) {
-    Optional<Principal> principal;
-    Optional<String> possibleClientName;
+    // Extract information about the requester. This may be a Keywhiz client, or it may be a proxy
+    // forwarding the real Keywhiz client information in the x-forwarded-client-certs header
+    Optional<Principal> requestPrincipal = getPrincipal(request);
+    if (requestPrincipal.isEmpty()) {
+      throw new NotAuthorizedException("Could not parse an identity from request");
+    }
+
+    Optional<String> requestName = getClientName(requestPrincipal);
+    Optional<String> requestSpiffeId = Optional.empty();
 
     // Extract client information based on the x-forwarded-client-cert header (if set) or
     // on the security context of this request
-    if (request.getRequestHeader(XFCC_HEADER_NAME) == null ||
-        request.getRequestHeader(XFCC_HEADER_NAME).isEmpty()) {
+    List<String> xfccHeaderValues = request.getRequestHeader(XFCC_HEADER_NAME);
+    if (xfccHeaderValues == null || xfccHeaderValues.isEmpty()) {
       // The XFCC header was not set; use the security context of this request to identify the client
```

Comment:
Falling back to using the mTLS client cert identity when XFCC is not present or is malformed might be unsafe. For example, let's say I can present to Envoy a client certificate it accepts and then forwards via XFCC, but then Keywhiz can't extract the caller's identity from XFCC. In that situation, Keywhiz will attempt to handle the request as though it originated from the mTLS identity Envoy is running as. If that mTLS identity on its own is granted non-trivial access in Keywhiz, then ....

I wonder whether it may be cleaner/safer to have a per-port setting of whether caller identity comes from XFCC or from the mTLS client certificate.

Reply:
Good point. So the configuration would look something like:
```
xfcc:
  - port: 1111
    allowedClientNames: [xfccClient1, xfccClient2]
  - port:2222
    allowedClientNames: [xfccClient3]
```
Requiring a different client list per port slightly increases complexity, but avoids another set of possible misconfigurations.

Then the workflow for handling an incoming request becomes:

1. Check the port that the request comes from and whether the XFCC header is set. 
2. Reject requests from "xfcc" ports that _don't_ have the header set and requests from non-"xfcc" ports that _do_ have the header set. (Either case means something is probably misconfigured.)
3. (already implemented, except for making it port specific:) Parse the mTLS cert for a client identity for a non-"xfcc" port/a sender identity for an "xfcc" port and check that identity against the allowed clients for that port. Reject requests from clients that aren't allowed to set XFCC for that port; otherwise parse the client identity out of the XFCC header.
4. (already implemented) Use that client identity to handle the request.

Reply:
@klyubin Updated--the configuration now specifically designates ports as "for x-forwarded-client-cert headers," and rejects both non-XFCC traffic sent to those ports and XFCC traffic sent to other ports.